### PR TITLE
Project Detail Page changes for Labels, links and autocomplete ajax.

### LINKED ISSUE #15577: Incorrect language string used on projects overview section list for Point Triggers
### Mautic Series

7.0.x series

### Mautic installed version

7.0

### Way of installing

I cloned the Mautic repo (not recommended or supported)

### PHP version

8.4

### What browsers are you seeing the problem on?

Not relevant

### What happened?

When I am viewing a project, the entities should be named according to the language string used in Mautic. It seems most are, but instead of Point Triggers being two words, it's one word:

Image

### How can we reproduce this issue?

Step 1: Create various resources including point triggers in Mautic and add them to a project
Step 2: Check the project overview and notice the point triggers heading isn't correctly formatted.

### Relevant log output

```shell

```

### Code of Conduct

- [x] I confirm that I have read and agree to follow this project's Code of Conduct

<br /><hr>
Care about this issue? Want to get it resolved sooner? If you are a <a href='https://www.mautic.org/become-a-member-of-mautic'>member of Mautic</a>, you can add some funds to the <a href='https://opencollective.com/mautic/projects/bounties'>Bounties Project</a> so that the person who completes this task can claim those funds once it is merged by a member of the core team! Read the docs <a href='https://contribute.mautic.org/product-team/mautic-bounty-programme'>here.</a>

### PR DESCRIPTION
**Pull request by dhirendramishra10**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | ✔️❌ 
| New feature/enhancement? (use the a.x branch)      | ✔️❌
| Deprecations?                          | ✔️❌
| BC breaks? (use the c.x branch)        | ✔️❌
| Automated tests included?              | ✔️❌ 
| Related user documentation PR URL      | mautic/user-documentation#... 
| Related developer documentation PR URL | mautic/developer-documentation-new#... 
| Issue(s) addressed                     | Fixes #... 

## Description

- Corrected wrong linking for Point and Stage
- Added Same name for each entity type as left hand side menu with plural
- Fixed issue for Modal default Autocomplete dropdown for entity types shown on Project details page
- Also fixed Modal Autocomplete dropdown for entity types shown on Project details page when user starts typing...

---
### 📋 Steps to test this PR:
 - Create one Project => lets say Project1
 - Create one more Project => lets say Project2
 - Create few landing pages (L1, L2, L3) , Points (Pn1, Pn2, Pn3) and Focus items etc
 - Now tag L1 with Project1 and L2, L3 with Project2
 - Now open pop up modal to add Landing page on Project1 detail page and you will see wrong landing pages are show and when typing then it shows data not found,
 - Also check url for Point and Stages and global token is corrected now on project detail page

Note : For Example:  check if listed entity items are landing to their landing page . for ex. point should land to /points/edit/{id} page

### CODE REVIEW

**Comment by dhirendramishra10** [comment]
Screenshot tested on local: 

Screenshot 2025-09-18 at 12 26 42 PM
Screenshot 2025-09-18 at 12 26 47 PM
Screenshot 2025-09-18 at 12 26 55 PM

**Comment by dhirendramishra10** [comment]
Community PR : https://github.com/mautic/mautic/pull/15493

**Comment by escopecz** [comment]
@dhirendramishra10 is this PR still WIP?

**Comment by dhirendramishra10** [comment]
@escopecz this is completed and tested on Local. there was just php-cs issue, now it is fixed.

**Comment by escopecz** [comment]
@dhirendramishra10 Then I'm removing the WIP from the title.

**Review by dadarya0 — APPROVED** [review]

File: app/bundles/ProjectBundle/Resources/views/Project/_entity_list.html.twig [hunk 1]
```diff
@@ -38,31 +38,44 @@
...
+												'objectAction': objectAction,
+												'objectId': entity.getId()
+										} %}
+
+										{% if entityType == 'global_token' %}
+												{# Special case: different route, no objectAction param #}
+												{% set route = 'mautic_global_token_edit' %}
+												{% set parameters = { 'objectId': entity.getId() } %}
+										{% endif %}
+
```
Comment (fedys) [inline app/bundles/ProjectBundle/Resources/views/Project/_entity_list.html.twig]: We cannot add conditions that create dependencies on other entities. Especially on those that even don't exist in the core bundles (e.g. `global_token`). We can address this by introducing an `entityData.detailRoute` value object.

```php
<?php

declare(strict_types=1);

namespace Mautic\ProjectBundle\DTO;


final readonly class EntityTypeConfig
{
    public function __construct(
        public string $entityClass,
        public string $label,
        public ?object $model = null,
        public ?DetailRoute $detailRoute = null,
    ) {
    }
}
```

```php
<?php

declare(strict_types=1);

namespace Mautic\ProjectBundle\DTO;

final readonly class DetailRoute
{
    public function __construct(
        public string $route,
        private string $idParameterName,
        private array $otherParameters = [],
    ) {
    }

    public function getParameters(string $id): array
    {
        return [$this->idParameterName => $id] + $this->otherParameters;
    }
}
```
We can set the `DetailRoute` objects via the existing or a new event (with default implementation) at https://github.com/mautic/mautic/blob/60b5e5deb1324e87f89fd154f36e3e6c53b7e630/app/bundles/ProjectBundle/Service/ProjectEntityLoaderService.php#L199

File: app/bundles/ProjectBundle/Resources/views/Project/_entity_list.html.twig [hunk 2]
```diff
@@ -38,31 +38,44 @@
...
 										<td>
-											<a href="{{ path('mautic_' ~ entityType ~ '_action', {'objectAction': 'view', 'objectId': entity.getId()}) }}" data-toggle="ajax">
-												{{ entity.getName()|purify }}
-											</a>
+												<a href="{{ path(route, parameters) }}" data-toggle="ajax">
```
Comment (fedys) [inline app/bundles/ProjectBundle/Resources/views/Project/_entity_list.html.twig]: Once we have the `detailRoute` object we can rewrite this line as follows:
```suggestion
												<a href="{{ path(entityData.detailRoute.route, entityData.detailRoute.getParameters(entity.getId())) }}" data-toggle="ajax">
```

File: app/bundles/ProjectBundle/Translations/en_US/messages.ini [hunk 3]
```diff
@@ -39,3 +39,16 @@ mautic.project.available_entity_types="Available Entity Types"
...
+mautic.project.asset = "Assets"
+mautic.project.pointtrigger = "Point Triggers"
+mautic.project.global_token = "Global Tokens"
+mautic.project.message = "Marketing Messages"
+mautic.project.form = "Forms"
+mautic.project.email = "Emails"
+mautic.project.stage = "Stages"
+mautic.project.segment = "Segments"
+mautic.project.company = "Companies"
+mautic.project.campaign = "Campaigns"
```
Comment (fedys) [inline app/bundles/ProjectBundle/Translations/en_US/messages.ini]: Again, this creates dependencies on other bundles. Let's move all of these to given bundles keeping the same ini keys.

**Review by fedys — CHANGES_REQUESTED** [review]

File: app/bundles/ProjectBundle/Resources/views/Project/_entity_list.html.twig (hunk 1 above)
Comment (dhirendramishra10) [reply app/bundles/ProjectBundle/Resources/views/Project/_entity_list.html.twig]: This is done

File: app/bundles/ProjectBundle/Resources/views/Project/_entity_list.html.twig (hunk 2 above)
Comment (dhirendramishra10) [reply app/bundles/ProjectBundle/Resources/views/Project/_entity_list.html.twig]: This is also done

File: app/bundles/ProjectBundle/Translations/en_US/messages.ini (hunk 3 above)
Comment (dhirendramishra10) [reply app/bundles/ProjectBundle/Translations/en_US/messages.ini]: This is also done

File: app/bundles/ProjectBundle/Service/ProjectEntityLoaderService.php [hunk 4]
```diff
@@ -51,9 +52,11 @@ public function getProjectEntities(Project $project, array $entityTypes): array
...
 
             $results[$entityType] = [
-                'label'    => $config->label,
-                'entities' => $entities,
-                'count'    => count($entities),
+                'label'       => $config->label,
+                'entities'    => $entities,
+                'count'       => count($entities),
+                'detailRoute' => $config->detailRoute
+        ?? new DetailRoute('mautic_'.$entityType.'_action', 'objectId', ['objectAction' => 'view']),
```
Comment (fedys) [inline app/bundles/ProjectBundle/Service/ProjectEntityLoaderService.php]:
```suggestion
                'detailRoute' => $config->detailRoute,
```
Let's not use a default detail route if it is not set. In theory in the future there might be an entity without detail page. Moreover this is a duplication of the [line number 211](https://github.com/mautic/mautic/pull/15493/files#diff-7581476e9d125cdd9761d1b4517a96fea4c558dcf994afca1b4ee4239c37fe56R211).

File: app/bundles/ProjectBundle/Service/ProjectEntityLoaderService.php [hunk 5]
```diff
@@ -181,6 +185,14 @@ private function getEntityTypes(): array
 
         $allMetadata = $this->em->getMetadataFactory()->getAllMetadata();
 
+        // Example mapping; only register what actually exists / is permitted.
+        $detailRoutes = [
+            'point'   => new DetailRoute('mautic_point_action', 'objectId', ['objectAction' => 'edit']),
+            'stage'   => new DetailRoute('mautic_stage_action', 'objectId', ['objectAction' => 'edit']),
+            'segment' => new DetailRoute('mautic_segment_action', 'objectId', ['objectAction' => 'view']),
+            // add more types as needed; do not add types for bundles you don't ship
+        ];
```
Comment (fedys) [inline app/bundles/ProjectBundle/Service/ProjectEntityLoaderService.php]:
```suggestion
        $routeEvent = new EntityTypeDetailRouteEvent();
        $this->eventDispatcher->dispatch($routeEvent);
```
The point of [my suggestion](https://github.com/mautic/mautic/pull/15493#discussion_r2426208726) was to remove the direct bidirectional coupling/dependency between projects and other entities. You just moved that dependency from the view layer to the model layer. Let's use events to address that.

The `EntityTypeDetailRouteEvent` might look like:

```php
<?php

declare(strict_types=1);

namespace Mautic\ProjectBundle\Event;

use Mautic\ProjectBundle\DTO\DetailRoute;
use Symfony\Contracts\EventDispatcher\Event;

final class EntityTypeDetailRouteEvent extends Event
{
    /**
     * @var array<?DetailRoute>
     */
    private array $routes = [];

    public function addRoute(string $entityType, ?DetailRoute $route): void
    {
        $this->routes[$entityType] = $route;
    }

    public function getRoute(string $entityType): ?DetailRoute
    {
        if (array_key_exists($entityType, $this->routes)) {
            return $this->routes[$entityType];
        }

        return new DetailRoute('mautic_'.$entityType.'_action', 'objectId', ['objectAction' => 'view']);
    }
}

```

File: app/bundles/ProjectBundle/Service/ProjectEntityLoaderService.php [hunk 6]
```diff
@@ -196,6 +208,7 @@ private function getEntityTypes(): array
                         entityClass: $entityClass,
                         label: $this->getEntityLabel($entityType),
                         model: $this->findModelForEntityType($entityType),
+                        detailRoute: $detailRoutes[$entityType] ?? new DetailRoute('mautic_'.$entityType.'_action', 'objectId', ['objectAction' => 'view']),
```
Comment (fedys) [inline app/bundles/ProjectBundle/Service/ProjectEntityLoaderService.php]:
```suggestion
                        detailRoute: $routeEvent->getRoute($entityType),
```
See my previous comment.

File: app/bundles/ProjectBundle/Resources/views/Project/_entity_list.html.twig [hunk 7]
```diff
@@ -38,30 +38,30 @@
...
+																								'priority': 100
+																						}
+																				},
+																		}) -}}
 										</td>
 										<td>
-											<a href="{{ path('mautic_' ~ entityType ~ '_action', {'objectAction': 'view', 'objectId': entity.getId()}) }}" data-toggle="ajax">
-												{{ entity.getName()|purify }}
+											<a href="{{ path(entityData.detailRoute.route, entityData.detailRoute.getParameters(entity.getId())) }}" data-toggle="ajax">
+													{{ entity.getName()|purify }}
```
Comment (fedys) [inline app/bundles/ProjectBundle/Resources/views/Project/_entity_list.html.twig]:
```suggestion
                                            {% if entityData.detailRoute %}
                                                <a href="{{ path(entityData.detailRoute.route, entityData.detailRoute.getParameters(entity.getId())) }}" data-toggle="ajax">
                                                    {{ entity.getName()|purify }}
                                                </a>
                                            {% else %}
                                                {{ entity.getName()|purify }}
                                            {% endif %}
```
Let's support entities that don't have any detail page.

**Review by fedys — CHANGES_REQUESTED** [review]

File: app/bundles/ProjectBundle/Service/ProjectEntityLoaderService.php [hunk 8]
```diff
@@ -181,6 +185,13 @@ private function getEntityTypes(): array
 
         $allMetadata = $this->em->getMetadataFactory()->getAllMetadata();
 
+        // Example mapping; only register what actually exists / is permitted.
+        $routeEvent = new EntityTypeDetailRouteEvent();
+        $this->eventDispatcher->dispatch($routeEvent);
+        $routeEvent->addRoute('point', new DetailRoute('mautic_point_action', 'objectId', ['objectAction' => 'edit']));
+        $routeEvent->addRoute('stage', new DetailRoute('mautic_stage_action', 'objectId', ['objectAction' => 'edit']));
+        $routeEvent->addRoute('segment', new DetailRoute('mautic_segment_action', 'objectId', ['objectAction' => 'view']));
```
Comment (fedys) [inline app/bundles/ProjectBundle/Service/ProjectEntityLoaderService.php]:
```suggestion
```
These details need to be provided by event listeners in given bundles.

**Review by fedys — CHANGES_REQUESTED** [review]

File: app/bundles/ProjectBundle/Service/ProjectEntityLoaderService.php (hunk 8 above)
Comment (dhirendramishra10) [reply app/bundles/ProjectBundle/Service/ProjectEntityLoaderService.php]: @fedys  This is addressed through event listeners in given bundles. For Segment it was not required to removed it.

Comment (fedys) [reply app/bundles/ProjectBundle/Service/ProjectEntityLoaderService.php]: Thank you!

**Review by fedys — APPROVED** [review]
The code changes work as expected.
Screenshot 2025-11-06 at 8 34 49
Screenshot 2025-11-06 at 8 38 00
Screenshot 2025-11-06 at 8 40 05
Screenshot 2025-11-06 at 9 16 26
